### ISSUE TITLE
[Feature Request] add WHEN NOT MATCHED BY SOURCE/TARGET clause suppoort

### ISSUE THREAD

**Issue by geoHeil** [delta-io/delta#1364]
## Feature request

https://delta-users.slack.com/archives/CJ70UCSHM/p1661955032288519

### Overview

WHEN NOT MATCHED BY SOURCE/TARGET clause support

### Motivation

feature parity with popular other SQL databases, ease of use

### Further details

Each day I get a full dump of a table. However, this data needs to be cleaned and in particular, compressed using the SCD2 style approach to be easily consumable downstream.
Unfortunately, I do not get changesets or a NULL value for a key in case of deletions. I only receive NO LONGER a row (including the key).
The links:
- https://docs.databricks.com/_static/notebooks/merge-in-scd-type-2.html
- https://docs.delta.io/latest/delta-update.html#write-change-data-into-a-delta-table
- https://docs.databricks.com/spark/latest/spark-sql/language-manual/delta-merge-into.html
show me that something along these lines (https://www.mssqltips.com/sqlservertip/1704/using-merge-in-sql-server-to-insert-update-and-delete-at-the-same-time):

```
MERGE <target_table> [AS TARGET]
USING <table_source> [AS SOURCE]
ON <search_condition>
[WHEN MATCHED 
   THEN <merge_matched> ]
[WHEN NOT MATCHED [BY TARGET]
   THEN <merge_not_matched> ]
[WHEN NOT MATCHED BY SOURCE
   THEN <merge_matched> ];

```
will not work with Delta/Spark as the WHEN NOT MATCHED clause does not seem to support the BY SOURCE | TARGET extension.
How can I still calculate the SCD2 representation?

1. If the table is empty simply take all the data (for the initial load)
2. When a new day/full copy of the data arrives:
  - INSERT any new keys into the table
  - For EXISTING keys perform an update (set the OLD value to be no longer valid (set end-date) and produce a new row in SCD2 with the contents of the new row and validity until infinity (end-date null))
  - In case a previously present key Is no longer present close the SCD2 valid_from/valid_to interval by setting end-date
  	- In case a new record arrives for this key in the future start a new fresh SCD2 row valid until infinity for this new row/values.

An example case/dataset:

```
import pandas as pd
import numpy as np
# assumes a running spark session including support for deltalake to be available

d1 = spark.createDataFrame(pd.DataFrame({'key':[1,2,3], 'value':[4,5,6],'value2':["a", "b", "c"], 'date':[1,1,1]}))
#d1.show()

# notice one entry is MISSING (it should be deleted) or rather SCD2 invalidated
d2 = spark.createDataFrame(pd.DataFrame({'key':[1,2], 'value':[4,5], 'date':[2,2],'value2':["a", "b"]}))

# d2 had (3) as missing - this entry is back now (and should start a new SCD2 row
d3 = spark.createDataFrame(pd.DataFrame({'key':[1,2,3], 'value':[4,5, 66], 'date':[3,3,3], 'value2':["a", "b", "c"]}))

# a new record is added
d4 = spark.createDataFrame(pd.DataFrame({'key':[1,2,3, 4], 'value':[4,5, 66, 44], 'date':[4,4,4,4], 'value2':["a", "b", "c", "d"]}))

# a new record is added, one removed and one updated
d5 = spark.createDataFrame(pd.DataFrame({'key':[2,3, 4, 5], 'value':[5, 67, 44, 55], 'date':[5,5,5,5], 'value2':["b", "c", "d", "e"]}))

```

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [ ] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [x] No. I cannot contribute this feature at this time.

**Comment by geoHeil** [comment]
Is there any workaround / current solution (perhaps not in the SQL but scala/Python API?

**Comment by nkarpov** [comment]
Thanks for creating this issue @geoHeil! Copying over a suggested approach from the Slack thread:
- No out-of-the-box way to do this in a single atomic command today
- You can LEFT join the target to the source and derive a new column for rows to DELETE, then run a MERGE with multiple WHEN MATCHED clauses to cover all the cases

Otherwise, the feature request is a good one, so we'll leave it open here if anyone from the community would like to work on it.

**Comment by geoHeil** [comment]
A small example:

step 1: initial write

```
# state cleanup
!rm -rf dummy_delta

# initial first write of data (first load of data)"
d1.write.mode("overwrite").format("delta").save("dummy_delta")
delta_state = spark.read.format("delta").load("dummy_delta")
delta_state.printSchema()
delta_state.show()
```

First UPSERT (including SCD2 soft delete):

```
# JOIN the existing state with the data

# DELETE
current_dt = 1.0
print(delta_state.count())
r = delta_state.join(d2.withColumnRenamed("valid_to", "valid_to_c").withColumnRenamed("value", "value_c").withColumnRenamed("date", "date_c"), on=['key'], how='left').withColumn("valid_to", F.when(F.col("value_c").isNull(), F.lit(current_dt)))
# shrink back (i.e. manual MERGE INTO)
r = r.select("key", "value", "date", "valid_to")
print(r.count())
r.show()

# SCD2 not relevant: in this example no additional rows/update rows are present 


# overwrite the latest state
r.write.mode("overwrite").format("delta").save("dummy_delta")
```


Next (2nd) UPSERT (including SCD2 soft delete):

> d2 had (3) as missing - this entry is back now (and should start a new SCD2 row with fresh valid_from valid until infinity (null for valid_to)

```
delta_state = spark.read.format("delta").load("dummy_delta")
current_dt = 2.0

r = delta_state.join(d3.withColumnRenamed("valid_to", "valid_to_c").withColumnRenamed("value", "value_c").withColumnRenamed("date", "date_c"), on=['key'], how='left').withColumn("valid_to", F.when(F.col("value_c").isNull(), F.lit(current_dt)))
# shrink back (i.e. manual MERGE INTO)
#r = r.select("key", "value", "date", "valid_to")
r.show()
```

questions: 

- how to handle multiple columns more efficiently this is super clumsy.
- how to turn a row into SCD2 (invalidate soft delete by setting valid_to and opening a new one with valid_from)? I can currently in this manual approach only think about explode(array(struct)) but this again feels super inefficient (a lot of shuffling involved). There must be a better approach

**Comment by geoHeil** [comment]
Some of the questions (around easing the handling of deletions) can be partially handled with:

```
def perform_rename(df, mapping):
    for old_value, new_value in mapping.items():
        df = df.withColumnRenamed(old_value, new_value)
    return df


def handle_deletions(updates_df, keys, current_dt, date_col='date',columns_to_drop=[], is_current_col='is_current', valid_to_col='valid_to'):
    def _(df_current_state):
        original_columns = delta_state.columns        
        mapping = dict([(c, f'{c}_c') for c in updates_df.drop(*keys).drop(*columns_to_drop).columns])
        value_cols = updates_df.drop(*keys).drop(*columns_to_drop).columns
        r = df_current_state.join(perform_rename(updates_df, mapping), on=keys, how='left')
        
        # it is good enough to look only at a single value (all wil be NULL anyways)
        v = f"{value_cols[0]}_c"
        r = r.withColumn(is_current_col, F.when(F.col(v).isNull(), F.lit(False)).otherwise(F.col(is_current_col))).withColumn(valid_to_col, F.when(F.col(v).isNull(), F.lit(current_dt)).otherwise(F.col(valid_to_col)))
        return r.select(original_columns)

    return _

current_dt = 1
deletions_handled = delta_state.transform(handle_deletions(d2, keys=['key'], current_dt=current_dt, columns_to_drop=['valid_to']))
deletions_handled.show()
```

**Comment by scottsand-db** [comment]
This feature seems neat! Unfortunately, we are already fully booked in our [H2 roadmap](https://github.com/delta-io/delta/issues/1307), and since you've indicated you are unable to contribute a solution, there's not much we can do right now.

When we begin our 2023 roadmap planning, please bring up this issue again so that we can get the community's input on how desired it is!

Cheers.

**Comment by seddonm1** [comment]
FYI I actually did all the work a couple of years ago and have a branch with this implemented for the Scala API only here: https://github.com/tripl-ai/delta

At the time the PR was rejected to this repo but if you are motivated the code could be updated for latest Delta (not by me).

**Comment by johanl-db** [comment]
I created a design doc to implement support for WHEN NOT MATCHED BY SOURCE clauses: [[Design Doc] WHEN NOT MATCHED BY SOURCE](https://docs.google.com/document/d/13DRQNu3k0ndUIvzGnNnyyiF9e5sQR1NurFVJxbhGgXY/edit?usp=sharing). This enables selectively updating or deleting target rows that have no matches in the source table based on the merge condition.

### API
A new `whenNotMatchedBySource(condition)` method is added to the Delta Table API, similar to the existing `whenMatched(condition)` and `whenNotMatched(condition)` methods.  It returns a builder that allows specifying the action to apply using `update(set)` or `delete()`. `whenNotMatchedBySource(condition)` accepts an optional condition that needs to be satisfied for the corresponding action to be applied.

Usage example:
```
targetDeltaTable.as(“t”)
  .merge(sourceTable.as(“s”), condition = “t.key = s.key”)
  .whenMatched().updateAll()
  .whenNotMatched().insertAll()
  .whenNotMatchedBySource(condition = “t.value > 0”).update(set = “t.value = 0”)
  .whenNotMatchedBySource().delete()
```
This merge invocation will:
- update all target rows that have a match in the source table using the source value.
- Insert all source rows that have no match in the target into the target table.
- Update all target rows that have no match in the source if `t.value` is strictly positive, otherwise delete the target row.

More details on the API and the implementation proposal can be found in the design doc. The SQL API will be shipped with Spark 3.4, see https://github.com/apache/spark/pull/38400.

**Comment by geoHeil** [comment]
solved by https://github.com/delta-io/delta/pull/1511

**Comment by scottsand-db** [comment]
Just posting for visibility, from #1511: `Support for WHEN NOT MATCHED BY SOURCE using SQL will be available with Spark 3.4 release and python support will follow up in a different PR.`

**Comment by allisonport-db** [comment]
SQL support is in https://github.com/delta-io/delta/pull/1740
